### PR TITLE
 Moving publisher from/to Sherlock.

### PR DESCRIPTION
**Pull request by mzhurovich**

### CODE REVIEW

File: Blocks/SS/test.cc [hunk 1]
```diff
@@ -77,15 +77,19 @@ TEST(StreamSystem, TestHelperClassSmokeTest) {
 namespace stream_system_test {
 
 struct DemoEntryPublisherImpl {
+  using MutexLockStatus = current::locks::MutexLockStatus;
+
   size_t index = 0u;
   std::deque<DispatchDemoEntry> values;  // Use `deque` to avoid extra copies.
 
+  template <MutexLockStatus MLS>
```
Comment (dkorolev) [inline Blocks/SS/test.cc]: Unused template parameters can be unnames, `template <MutexLockStatus>`. Another nit: If a fully-qualified name fits the line, it's better for test readability and copy-and-paste-ability, ex. `template <current::locks::MutexLockStatus>`.

File: Bricks/util/locks.h [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+struct NoOpLock {
+  template <typename... ARGS>
+  NoOpLock(ARGS&&...) {}
+};
+
+template <MutexLockStatus MLS, class MUTEX = std::mutex>
+using SmartMutexLockGuard =
+    typename std::conditional<MLS == MutexLockStatus::NeedToLock, std::lock_guard<MUTEX>, NoOpLock>::type;
+
+static_assert(std::is_same<std::lock_guard<std::mutex>, SmartMutexLockGuard<MutexLockStatus::NeedToLock>>::value, "");
```
Comment (dkorolev) [inline Bricks/util/locks.h]: :+1:

File: Bricks/util/locks.h [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+static_assert(std::is_same<NoOpLock, SmartMutexLockGuard<MutexLockStatus::AlreadyLocked>>::value, "");
+
+}  // namespace locks
+}  // namespace current
+
```
Comment (dkorolev) [inline Bricks/util/locks.h]: Redundant vertical whitespace.

File: Sherlock/sherlock.h [hunk 4]
```diff
@@ -134,30 +136,46 @@ class StreamImpl {
         : persistence(std::forward<ARGS>(args)...) {}
   };
 
+  class StreamPublisher {
+   public:
+    explicit StreamPublisher(std::shared_ptr<StreamData> data) : data_(data) {}
+
+    template <MutexLockStatus MLS>
```
Comment (dkorolev) [inline Sherlock/sherlock.h]: Full `enum class` name instead of `using` is probably cleaner here as well.

File: Sherlock/sherlock.h [hunk 5]
```diff
@@ -396,6 +451,8 @@ class StreamImpl {
 
  private:
   std::shared_ptr<StreamData> data_;
+  std::unique_ptr<publisher_t> publisher_;
```
Comment (dkorolev) [inline Sherlock/sherlock.h]: :+1:

File: Sherlock/test.cc [hunk 6]
```diff
@@ -665,3 +665,65 @@ TEST(Sherlock, SubscribeWithFilterByType) {
     EXPECT_EQ("Y=2 Y=4", Join(c.results_, ' '));
   }
 }
+
+namespace sherlock_unittest {}
```
Comment (dkorolev) [inline Sherlock/test.cc]: vim dd ?

File: Sherlock/test.cc [hunk 7]
```diff
@@ -665,3 +665,65 @@ TEST(Sherlock, SubscribeWithFilterByType) {
...
+
+TEST(Sherlock, ReleaseAndAcquirePublisher) {
+  using namespace sherlock_unittest;
```
Comment (dkorolev) [inline Sherlock/test.cc]: Ditto.

Comment (mzhurovich) [inline Sherlock/test.cc]: `Record` & `SherlockTestProcessor` are defined there.

**Pull request merged**